forward/forward.go

```diff
@@ -66,3 +68,79 @@ func Wrap(msg stanza.Message, body string, received time.Time, r xml.TokenReader
 		}.Wrap(r),
 	))
 }
+
+type forwardUnwrapper struct {
+	r         xml.TokenReader
+	stanzaXML xml.TokenReader
+	del       *delay.Delay
+}
+
+// Token implements xml.TokenReader
+func (f *forwardUnwrapper) Token() (xml.Token, error) {
+	if f.stanzaXML != nil {
+		token, err := f.stanzaXML.Token()
+		if err != nil && err != io.EOF {
+			// something bad happened
+			return nil, err
+		}
+		if err == io.EOF {
+			// we have reached the end of the stanza xml stream
+			f.stanzaXML = nil
+		}
+		if token != nil {
+			return token, nil
+		}
+	}
+
+	for {
+		token, err := f.r.Token()
+		if err != nil {
+			return nil, err
+		}
+		se, ok := token.(xml.StartElement)
+		if !ok {
+			return se, fmt.Errorf("expected a startElement, found %T", token)
+		}
+
+		if se.Name.Local == "delay" && se.Name.Space == delay.NS {
+			if f.del == nil {
+				// The delay element is skipped if the provided del is nil
+				err = xmlstream.Skip(f.r)
+				if err != nil {
+					return nil, err
+				}
+				continue
+			}
+			err = xml.NewTokenDecoder(
+				xmlstream.MultiReader(xmlstream.Token(se), xmlstream.Inner(f.r), xmlstream.Token(se.End())),
```

Comment:
I think this line can go away like I mentioned in another comment, but just for future reference this is the same as `xmlstream.Wrap(xmlstream.Inner(f.r), se)`